src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,458 @@
+            scalar: ecc_config.advices[1],
+            perm: ecc_config.perm.clone(),
+            add_config: ecc_config.into(),
+            hi_config: incomplete::Config::hi_config(ecc_config),
+            lo_config: incomplete::Config::lo_config(ecc_config),
+        };
+
+        assert_eq!(
+            config.hi_config.x_p, config.lo_config.x_p,
+            "x_p is shared across hi and lo halves."
+        );
+        assert_eq!(
+            config.hi_config.y_p, config.lo_config.y_p,
+            "y_p is shared across hi and lo halves."
+        );
+
+        let add_config_advices = config.add_config.advice_columns();
+        assert!(
+            !add_config_advices.contains(&config.z_complete),
+            "z_complete cannot overlap with complete addition columns."
+        );
+        assert!(
+            !add_config_advices.contains(&config.hi_config.z),
+            "hi_config z cannot overlap with complete addition columns."
+        );
+
+        config
+    }
+}
+
+impl<C: CurveAffine> Config<C>
+where
+    C::Scalar: PrimeFieldBits,
+{
+    pub(super) fn create_gate(&self, meta: &mut ConstraintSystem<C::Base>) {
+        self.hi_config.create_gate(meta);
+        self.lo_config.create_gate(meta);
+
+        let complete_config: complete::Config<C> = self.into();
+        complete_config.create_gate(meta);
+
+        self.create_final_scalar_gate(meta);
+    }
+
+    /// Gate used to check final scalar is recovered.
+    fn create_final_scalar_gate(&self, meta: &mut ConstraintSystem<C::Base>) {
+        meta.create_gate("Decompose scalar for variable-base mul", |meta| {
+            let q_mul_decompose_var = meta.query_selector(self.q_mul_decompose_var);
+            let scalar = meta.query_advice(self.scalar, Rotation::cur());
+            let z_cur = meta.query_advice(self.z_complete, Rotation::cur());
+
+            // The scalar field `F_q = 2^254 + t_q`.
+            // -((2^127)^2) = -(2^254) = t_q (mod q)
+            let t_q = -(C::Scalar::from_u128(1u128 << 127).square());
+            let t_q = C::Base::from_bytes(&t_q.to_bytes()).unwrap();
+
+            // Check that `k = scalar + t_q`
+            vec![q_mul_decompose_var * (scalar + Expression::Constant(t_q) - z_cur)]
+        });
+    }
+
+    pub(super) fn assign_region(
+        &self,
+        scalar: &CellValue<C::Base>,
+        base: &EccPoint<C>,
+        offset: usize,
+        region: &mut Region<'_, C::Base>,
+    ) -> Result<EccPoint<C>, Error> {
+        // Initialize the accumulator `acc = [2]base`
+        let acc = self
+            .add_config
+            .assign_region(&base, &base, offset, region)?;
+
+        // Increase the offset by 1 after complete addition.
+        let offset = offset + 1;
+
+        // Decompose the scalar bitwise (big-endian bit order).
+        let bits = decompose_for_scalar_mul::<C>(scalar.value());
+
+        // Initialize the running sum for scalar decomposition to zero
+        let z_val = C::Base::zero();
```

Comment:
What gate is enforcing that this must be zero?

Reply:
I've added a gate checking this in ae62bb2672c3dd466a5d59da3264afa7a522cfb2. Additionally, we are already checking that:
1. The scalar is decomposed bitwise; and
2. We recover the original scalar using the final value of the running sum.

Reply:
Resolving this because the running sum initialisation gate has been implemented at https://github.com/zcash/orchard/pull/111/files#diff-8c85db55a23b99719dabdaf6f8ea93578fe5921af5fc3e3802985a8394fa4fb1R85.